File: SKILLS/s-niftypm/scripts/reverse-sync.py
#!/usr/bin/env python3
"""
reverse-sync.py - Build a local NiftyPM project plan JSON from live API data.

Reads a bundle file (assembled from niftypm-mcp responses) and produces a
schema-compliant JSON per s-niftypm/references/niftypm-project-plan-schema.md.

This is the complement to Rule 5 (JSON-First Planning). Rule 5 covers
JSON -> NiftyPM (populating a new project). This script covers
NiftyPM -> JSON (syncing a live project into local source-of-truth).

Usage:
    python3 reverse-sync.py \\
        --bundle /tmp/NMQ-bundle.json \\
        --output NiftyPM/ISPF/NeurMantiq.json

Bundle file shape (see references/reverse-sync-workflow.md for assembly):
    {
      "project":   { id, nice_id, name, description, portfolio, portfolio_id, repo },
      "labels":    [ ... from list_labels ... ],
      "taskgroups":[ ... from list_taskgroups ... ],
      "milestones":[ ... merge list_milestones + list_milestones(is_list=true) ... ],
      "tasks":     [ ... from list_tasks ... ],
      "members":   [ ... from list_members ... ]
    }
"""
import argparse
import json
import sys
from pathlib import Path
from datetime import datetime, timezone


def date_only(iso_str):
    """Extract YYYY-MM-DD from an ISO 8601 datetime string, or None."""
    if not iso_str or not isinstance(iso_str, str):
        return None
    return iso_str[:10] if len(iso_str) >= 10 else None


def build_label_map(labels):
    """ID -> (name, color)."""
    return {
        l["id"]: (l.get("name", ""), l.get("color", "#000000"))
        for l in labels
        if "id" in l
    }


def build_taskgroup_map(taskgroups):
    """ID -> (name, order)."""
    return {
        tg["id"]: (tg.get("name", ""), tg.get("order", 0))
        for tg in taskgroups
        if "id" in tg
    }


def build_milestone_map(milestones):
    """ID -> (name, due_date, description).

    Merges regular + list-type milestones (dedup by id).
    Skips 'Untitled List' default artefacts.
    """
    result = {}
    for ms in milestones:
        if "id" not in ms:
            continue
        name = ms.get("name", "")
        if name.strip().lower() == "untitled list":
            continue
        ms_id = ms["id"]
        if ms_id in result:
            continue
        due = date_only(ms.get("end")) or date_only(ms.get("start"))
        desc = ms.get("description") or ""
        result[ms_id] = (name, due, desc)
    return result


def build_project_json(bundle):
    """Transform a bundle into schema-compliant JSON."""
    now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

    project = bundle.get("project", {})
    tasks_raw = bundle.get("tasks", [])
    label_map = build_label_map(bundle.get("labels", []))
    tg_map = build_taskgroup_map(bundle.get("taskgroups", []))
    ms_map = build_milestone_map(bundle.get("milestones", []))

    id_to_nice = {
        t["id"]: t["nice_id"]
        for t in tasks_raw
        if "id" in t and "nice_id" in t
    }

    used_label_ids = set()
    for t in tasks_raw:
        used_label_ids.update(t.get("labels") or [])

    out_tasks = []
    for t in tasks_raw:
        tg_id = t.get("task_group")
        ms_id = t.get("milestone")
        dep_id = t.get("dependency")
        label_ids = t.get("labels") or []

        tg_name = tg_map.get(tg_id, ("Unknown",))[0] if tg_id else None
        ms_name = ms_map.get(ms_id, (None,))[0] if ms_id else None
        dep_nice = id_to_nice.get(dep_id) if dep_id in id_to_nice else None

        out_tasks.append({
            "id": t.get("id"),
            "nice_id": t.get("nice_id"),
            "name": t.get("name", ""),
            "task_list": tg_name,
            "milestone": ms_name,
            "description": t.get("description") or "",
            "labels": [label_map[l][0] for l in label_ids if l in label_map],
            "story_points": t.get("story_points"),
            "due_date": date_only(t.get("due_date")),
            "start_date": date_only(t.get("start_date")),
            "dependency": dep_nice,
            "subtasks": [],
            "total_subtasks": t.get("total_subtasks", 0),
            "completed": t.get("completed", False),
            "completed_on": t.get("completed_on"),
            "assignees": t.get("assignees") or [],
            "archived": t.get("archived", False),
        })

    completed_count = sum(1 for t in out_tasks if t["completed"])
    no_milestone = sum(1 for t in out_tasks if t["milestone"] is None)
    no_task_list = sum(1 for t in out_tasks if t["task_list"] is None)
    used_labels_resolved = [l for l in used_label_ids if l in label_map]

    result = {
        "meta": {
            "created": now,
            "last_synced": now,
            "generated_by": "s-niftypm/scripts/reverse-sync.py",
            "project_nice_id": project.get("nice_id"),
            "niftypm_project_id": project.get("id"),
            "source": "Live NiftyPM API via niftypm-mcp",
            "notes": (
                "Subtask names not yet populated (requires get_item_children "
                "per parent task). All other fields resolved from live data."
            ),
        },
        "project": {
            "name": project.get("name", ""),
            "description": project.get("description", ""),
            "portfolio": project.get("portfolio", ""),
            "portfolio_id": project.get("portfolio_id"),
            "repo": project.get("repo"),
        },
        "labels": [
            {"name": label_map[lid][0], "color": label_map[lid][1], "id": lid}
            for lid in sorted(used_labels_resolved)
        ],
        "task_lists": [
            {"name": name, "id": tid, "order": order}
            for tid, (name, order) in sorted(tg_map.items(), key=lambda x: x[1][1] or 0)
        ],
        "milestones": [
            {"name": ms[0], "id": mid, "due": ms[1], "description": ms[2]}
            for mid, ms in ms_map.items()
        ],
        "tasks": out_tasks,
        "_validation_checklist": [
            f"Synced from live NiftyPM API on {now}",
            f"Tasks: {len(out_tasks)} total ({completed_count} completed)",
            f"Labels used: {len(used_labels_resolved)}",
            f"Tasks without milestone: {no_milestone} (null is honest for retrospective done-records)",
            f"Tasks without task_list: {no_task_list}",
            "Subtask names are a known gap - enrich via get_item_children in a follow-up pass",
            "All task IDs, nice_ids, dependencies, and milestone assignments resolved from live data",
        ],
    }
    return result


def main():
    parser = argparse.ArgumentParser(
        description="Build a NiftyPM project plan JSON from a live API data bundle."
    )
    parser.add_argument(
        "--bundle", required=True,
        help="Path to bundle JSON file (assembled from niftypm-mcp responses)"
    )
    parser.add_argument(
        "--output", required=True,
        help="Path to output JSON file"
    )
    args = parser.parse_args()

    bundle_path = Path(args.bundle)
    output_path = Path(args.output)

    if not bundle_path.exists():
        print(f"ERROR: Bundle file not found: {bundle_path}", file=sys.stderr)
        sys.exit(1)

    with open(bundle_path, encoding="utf-8") as f:
        bundle = json.load(f)

    result = build_project_json(bundle)

    output_path.parent.mkdir(parents=True, exist_ok=True)
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(result, f, indent=2, ensure_ascii=False)

    tasks = result["tasks"]
    completed = sum(1 for t in tasks if t["completed"])
    print(
        f"WROTE {output_path.name}: "
        f"{len(tasks)} tasks ({completed} completed), "
        f"{len(result['labels'])} labels, "
        f"{len(result['milestones'])} milestones, "
        f"{output_path.stat().st_size:,} bytes"
    )


if __name__ == "__main__":
    main()

File: src/reverse-sync.ts
/**
 * TypeScript port of reverse-sync.py transform logic.
 * Shared by `init`, `sync`, and auto-sync.
 * Same output schema as the Python script.
 */

// ── Bundle input types ──────────────────────────────────────────────

export interface BundleProject {
  id?: string;
  nice_id?: string;
  name?: string;
  description?: string;
  portfolio?: string;
  portfolio_id?: string;
  repo?: string;
}

export interface BundleLabel {
  id: string;
  name?: string;
  color?: string;
}

export interface BundleTaskgroup {
  id: string;
  name?: string;
  order?: number;
}

export interface BundleMilestone {
  id: string;
  name?: string;
  end?: string;
  start?: string;
  description?: string;
}

export interface BundleTask {
  id: string;
  nice_id?: string;
  name?: string;
  task_group?: string;
  milestone?: string;
  dependency?: string;
  labels?: string[];
  description?: string;
  story_points?: number | null;
  due_date?: string | null;
  start_date?: string | null;
  total_subtasks?: number;
  completed?: boolean;
  completed_on?: string | null;
  assignees?: string[];
  archived?: boolean;
}

export interface BundleMember {
  id: string;
  name?: string;
  email?: string;
}

export interface Bundle {
  project?: BundleProject;
  labels?: BundleLabel[];
  taskgroups?: BundleTaskgroup[];
  milestones?: BundleMilestone[];
  tasks?: BundleTask[];
  members?: BundleMember[];
}

// ── Output types ────────────────────────────────────────────────────

export interface ProjectJsonMeta {
  created: string;
  last_synced: string;
  generated_by: string;
  project_nice_id?: string;
  niftypm_project_id?: string;
  source: string;
  notes: string;
}

export interface ProjectJsonProject {
  name: string;
  description: string;
  portfolio: string;
  portfolio_id?: string;
  repo?: string;
}

export interface ProjectJsonLabel {
  name: string;
  color: string;
  id: string;
}

export interface ProjectJsonTaskList {
  name: string;
  id: string;
  order: number;
}

export interface ProjectJsonMilestone {
  name: string;
  id: string;
  due: string | null;
  description: string;
}

export interface ProjectJsonTask {
  id: string;
  nice_id?: string;
  name: string;
  task_list: string | null;
  milestone: string | null;
  description: string;
  labels: string[];
  story_points: number | null;
  due_date: string | null;
  start_date: string | null;
  dependency: string | null;
  subtasks: any[];
  total_subtasks: number;
  completed: boolean;
  completed_on: string | null;
  assignees: string[];
  archived: boolean;
}

export interface ProjectJson {
  meta: ProjectJsonMeta;
  project: ProjectJsonProject;
  labels: ProjectJsonLabel[];
  task_lists: ProjectJsonTaskList[];
  milestones: ProjectJsonMilestone[];
  tasks: ProjectJsonTask[];
  _validation_checklist: string[];
}

// ── Helpers ─────────────────────────────────────────────────────────

function dateOnly(isoStr?: string | null): string | null {
  if (!isoStr || typeof isoStr !== "string") return null;
  return isoStr.length >= 10 ? isoStr.slice(0, 10) : null;
}

function buildLabelMap(labels: BundleLabel[]): Map<string, [string, string]> {
  const m = new Map<string, [string, string]>();
  for (const l of labels) {
    if (l.id) {
      m.set(l.id, [l.name || "", l.color || "#000000"]);
    }
  }
  return m;
}

function buildTaskgroupMap(taskgroups: BundleTaskgroup[]): Map<string, [string, number]> {
  const m = new Map<string, [string, number]>();
  for (const tg of taskgroups) {
    if (tg.id) {
      m.set(tg.id, [tg.name || "", tg.order || 0]);
    }
  }
  return m;
}

function buildMilestoneMap(
  milestones: BundleMilestone[],
): Map<string, [string, string | null, string]> {
  const m = new Map<string, [string, string | null, string]>();
  for (const ms of milestones) {
    if (!ms.id) continue;
    const name = ms.name || "";
    if (name.trim().toLowerCase() === "untitled list") continue;
    if (m.has(ms.id)) continue;
    const due = dateOnly(ms.end) || dateOnly(ms.start);
    const desc = ms.description || "";
    m.set(ms.id, [name, due, desc]);
  }
  return m;
}

// ── Main transform ──────────────────────────────────────────────────

export function buildProjectJson(bundle: Bundle): ProjectJson {
  const now = new Date().toISOString().replace(/\.\d+Z$/, "Z");

  const project = bundle.project || {};
  const tasksRaw = bundle.tasks || [];
  const labelMap = buildLabelMap(bundle.labels || []);
  const tgMap = buildTaskgroupMap(bundle.taskgroups || []);
  const msMap = buildMilestoneMap(bundle.milestones || []);

  const idToNice = new Map<string, string>();
  for (const t of tasksRaw) {
    if (t.id && t.nice_id) {
      idToNice.set(t.id, t.nice_id);
    }
  }

  const usedLabelIds = new Set<string>();
  for (const t of tasksRaw) {
    for (const lid of t.labels || []) {
      usedLabelIds.add(lid);
    }
  }

  const outTasks: ProjectJsonTask[] = tasksRaw.map((t) => {
    const tgId = t.task_group;
    const msId = t.milestone;
    const depId = t.dependency;
    const labelIds = t.labels || [];

    const tgName = tgId ? (tgMap.get(tgId)?.[0] ?? null) : null;
    const msName = msId ? (msMap.get(msId)?.[0] ?? null) : null;
    const depNice = depId && idToNice.has(depId) ? (idToNice.get(depId) ?? null) : null;

    return {
      id: t.id,
      nice_id: t.nice_id,
      name: t.name || "",
      task_list: tgName,
      milestone: msName,
      description: t.description || "",
      labels: labelIds.filter((l) => labelMap.has(l)).map((l) => labelMap.get(l)?.[0] ?? ""),
      story_points: t.story_points ?? null,
      due_date: dateOnly(t.due_date),
      start_date: dateOnly(t.start_date),
      dependency: depNice,
      subtasks: [],
      total_subtasks: t.total_subtasks || 0,
      completed: t.completed || false,
      completed_on: t.completed_on ?? null,
      assignees: t.assignees || [],
      archived: t.archived || false,
    };
  });

  const completedCount = outTasks.filter((t) => t.completed).length;
  const noMilestone = outTasks.filter((t) => t.milestone === null).length;
  const noTaskList = outTasks.filter((t) => t.task_list === null).length;
  const usedLabelsResolved = [...usedLabelIds].filter((l) => labelMap.has(l));

  return {
    meta: {
      created: now,
      last_synced: now,
      generated_by: "niftypm-mcp/reverse-sync.ts",
      project_nice_id: project.nice_id,
      niftypm_project_id: project.id,
      source: "Live NiftyPM API via niftypm-mcp",
      notes:
        "Subtask names not yet populated (requires get_item_children per parent task). All other fields resolved from live data.",
    },
    project: {
      name: project.name || "",
      description: project.description || "",
      portfolio: project.portfolio || "",
      portfolio_id: project.portfolio_id,
      repo: project.repo,
    },
    labels: [...usedLabelsResolved]
      .sort((a, b) => a.localeCompare(b))
      .map((lid) => {
        const entry = labelMap.get(lid);
        return { name: entry?.[0] ?? "", color: entry?.[1] ?? "", id: lid };
      }),
    task_lists: [...tgMap.entries()]
      .sort((a, b) => (a[1][1] || 0) - (b[1][1] || 0))
      .map(([tid, [name, order]]) => ({ name, id: tid, order })),
    milestones: [...msMap.entries()].map(([mid, [name, due, desc]]) => ({
      name,
      id: mid,
      due,
      description: desc,
    })),
    tasks: outTasks,
    _validation_checklist: [
      `Synced from live NiftyPM API on ${now}`,
      `Tasks: ${outTasks.length} total (${completedCount} completed)`,
      `Labels used: ${usedLabelsResolved.length}`,
      `Tasks without milestone: ${noMilestone} (null is honest for retrospective done-records)`,
      `Tasks without task_list: ${noTaskList}`,
      "Subtask names are a known gap - enrich via get_item_children in a follow-up pass",
      "All task IDs, nice_ids, dependencies, and milestone assignments resolved from live data",
    ],
  };
}
